### PR TITLE
Websocket gateway

### PR DESCRIPTION
**Pull request by jbjjbjjbj**
TODO:
  - [x] Fix Keyerror exception
  - [x] Return some exceptions back as errors on the websocket
  - [ ] Add documentation and a README

### CODE REVIEW

**Comment by jbjjbjjbj** [comment]
Okay i could maybe use some input here.
I'm currently mapping method calls and returns to json objects which are sent over the websocket.

So the status call will become:

```json
{"method": "Status", "value": {"queue_max": 1, "autoplay_max": 1}}
{
  "type": "Status.Info",
  "value": {
    "current": {
      "title": "",
      "artist": "",
      "album": ""
    },
    "length": 1,
    "progress": 0,
    "queue_length": 0,
    "queue_hash": 5740354900026072000,
    "queue": [],
    "autoplay": []
  }
}

```

Here the first object is the call message and the second is the response.

The way I'm doing this currently is not very automated, and the status mapping above can be seen below.
I'm not a big fan, but it's seems like the simplest solution.

```python
@api_method("Status", {"queue_max": int, "autoplay_max": int})
def status(sonica, socket, value):
    resp = sonica.Status(pb2.Status.Query(**value))
    return "Status.Info", dict_from_msg(resp, [
        ("current", dict_from_song),
        (["length", "progress", "queue_length", "queue_hash"], None),
        (["queue", "autoplay"], lambda x: [dict_from_song(s) for s in x])
    ])
```

Does this seem reasonable? Or should we try to find a more automatic way of mapping between grpc and websocket.

**Comment by SEbbaDK** [comment]
>```json
>{"method": "Status", "token": 0, "value": {"queue_max": 1, "autoplay_max": 1}}
>```

What is the token value here?

**Comment by SEbbaDK** [comment]
> Does this seem reasonable? Or should we try to find a more automatic way of mapping between grpc and websocket.

We've already tried with both OpenAPI and gRPC Gateway to map it automatically, but it hasn't worked so far, and the Sonica API is not super big, so manual mapping is fine for me unless we actively know some way to map it.

**Comment by jbjjbjjbj** [comment]
> > ```json
> > {"method": "Status", "token": 0, "value": {"queue_max": 1, "autoplay_max": 1}}
> > ```
> 
> What is the token value here?

The idea is that the client can choose a token when running a request.
The same token is then used in the response, such that the client knows which responses map to which requests.

**Comment by SEbbaDK** [comment]
> So i can call 10 commands at the same time, and still distinguish the results.

~~Is this not something that would normally be handled client-side? All calls to grpc are blocking, so i think the same idea goes for the api?~~
Ahh, this is a websocket and not a rest interface, gotcha!

File: gateway/sonica-gateway.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+                response = handle_message(message, sonica, websocket)
+                await websocket.send(json.dumps(response))
+            except:
+                await websocket.send(json.dumps({ "type" : "error", "value" : "Invalid JSON" }))
+
+    async with websockets.serve(echo, host, port):
+        await asyncio.Future()
+
+if __name__ == "__main__":
+    address = "localhost:7700"
```
Comment (SEbbaDK) [inline gateway/sonica-gateway.py:76]: This should be made into a CLI option

**Review by SEbbaDK — CHANGES_REQUESTED** [review]

**Review by SEbbaDK — APPROVED** [review]

**Pull request merged**